ballerina/vector_store.bal

```diff
@@ -0,0 +1,249 @@
+                        id: entry.id,
+                        values: embedding,
+                        metadata
+                    };
+                } else {
+                    return error ai:Error("Dense mode requires DenseVector embedding.");
+                }
+            } else if self.queryMode == ai:SPARSE {
+                if embedding is ai:SparseVector {
+                    vec = {
+                        id: entry.id,
+                        sparseValues: embedding,
+                        metadata
+                    };
+                } else {
+                    return error ai:Error("Sparse mode requires SparseVector embedding.");
+                }
+            } else if self.queryMode == ai:HYBRID {
+                if embedding is ai:HybridVector {
+                    if embedding.dense.length() == 0 && embedding.sparse.indices.length() == 0 {
+                        return error ai:Error("Hybrid mode requires both dense and sparse vectors, " +
+                        "but one or both are missing.");
+                    }
+                    vec = {
+                        id: entry.id,
+                        values: embedding.dense,
+                        sparseValues: embedding.sparse,
+                        metadata
+                    };
+                } else {
+                    return error ai:Error("Hybrid mode requires DenseVector and SparseVector embedding.");
+                }
+            } else {
+                return error ai:Error("Unsupported query mode.");
+            }
+
+            vectors.push(vec);
+        }
+
+        vector:UpsertRequest request = {
+            vectors: vectors
+        };
+
+        if self.namespace != "" {
+            request.namespace = self.namespace;
+        }
+
+        vector:UpsertResponse|error response = self.pineconeClient->/vectors/upsert.post(request);
+        if response is error {
+            log:printError("Failed to add vector entry", response);
+            return error ai:Error("Failed to add vector entry", response);
+        }
+    }
+
+    # Queries Pinecone using the provided embedding vector and returns the top matches.
+    #
+    # + queryVector - The embedding vector to query against. Should match the configured query mode
+    # + return - A list of matching ai:VectorMatch values, or an ai:Error on failure
+    public isolated function query(ai:VectorStoreQuery queryVector) returns ai:VectorMatch[]|ai:Error {
+        vector:QueryRequest request = {
+            topK: self.similarityTopK,
+            includeMetadata: true,
+            includeValues: true
+        };
+
+        if queryVector.embedding is ai:Vector {
+            if self.queryMode == ai:HYBRID {
+                return error ai:Error("Hybrid search requires both dense and sparse vectors, " +
+                "but only dense vector provided.");
+            }
+            request.vector = <ai:Vector>queryVector.embedding;
```

Comment:
let's introduce a variable and store `queryVector.embedding`. This way we can omit casting here.
Seems like there are few other places we do the casting let's update those places too.